muspinsim/spinop.py

```diff
@@ -329,16 +352,15 @@ def basis_change(self, basis):
         """
 
         ans = self.clone()
-        basis = sparse.csr_matrix(basis)
         x = basis.T.conjugate()
-        ans._matrix = x.dot(ans._matrix).dot(basis)
+        ans._matrix = x @ ans._matrix @ basis
```

Comment:
> This is my first time with the scipy sparse matrices, so please correct me if I'm misunderstanding at any point. Based on some experimenting, using `.dot` will preserve the object as a `csr_matrix`, whereas `@` will result in (implicitly) a dense numpy array iff basis is dense (not sure if this happens before or after calculation but that's besides the point).
> 

Yes so effectively if I used .dot, it doesn’t behave as you might expect and it ends up generating a list of csr_matrix objects rather than a single matrix. With relatively recent SciPy changes * has been replaced by @ for matrix multiplication, and it just so happens to mean the same thing in numpy so if both matrices are sparse it produces a sparse matrix, if first one is dense it produces a dense matrix. It looks like it internally converts to dense then multiplies based on memory usage, so it will use more memory overall, my test system above didn’t really show it – I forgot it was using the method that doesn’t even use `basis_change`. If I force it to use the method I get:

802MB, 4 minutes 35 seconds
Vs
703MB 2 minutes 52 seconds
By simply adding
```
field
   1
temperature
   1
```

This does beg the question, why did it show a performance improvement for my first change at all, equally why is the memory usage decreasing here… (despite trying it 4 times and consistently getting similar results) It seems something weird is going on. In any case it is better now than before.


> Most of the dunder methods on the operator preserve the sparsity (i.e. checking `_sparse` before using `@` in `__mul__`). All but one of the places we're using `basis_change` followed by `toarray` were changed as part of this PR anyway. So I'm (mostly) happy that this won't break anything...
> 
> https://github.com/muon-spectroscopy-computational-project/muspinsim/blob/a080b42580a1a93030d751dca0c86dfc01088553/muspinsim/hamiltonian.py#L112-L115
> 
> This usage is unchanged. Before, this would have meant creating a sparse matrix, and returning a sparse matrix from `basis_change`. But now won't this be a dense matrix since the eigenvectors come from `diag` which uses numpy so must be dense? I doesn't seem to be causing any tests to fail, so either:
> 
> 1. It's not covered by tests
> 2. It doesn't matter that it's sparse or dense and gets sorted out in the wash
> 3. I've completely misunderstood what's going on in the code
> 
> I would try and test further but figured I was already quite far down the rabbit hole so thought I'd ask you what you think before going any further?
> 
> Regardless of whether this is an issue or not, I do think that it might be worth adding something to the docstring to document that this will return a dense matrix regardless of whether the operator was sparse before the basis change. Since the other operations in the class will preserve it I'd be more comfortable if the one(?) place we don't is documented for the future.

Your right, it will now be dense. We don’t ever use evolve without an operator, so I don’t see why we have it apart from for usage as a library. It turns out it didn’t work anyway as it was broken when calc_single_rho was changed ~ 6 months ago. I would suggest we just remove it. I don’t think it makes sense to be transposing and getting the conjugate of the eigenvectors twice like this, which has been done since it was added in https://github.com/muon-spectroscopy-computational-project/muspinsim/commit/423cb722f2710201264da9e31e7295bac83a6d4a. But it is also simple to fix if we would rather keep it.

I was going to do one more release with the fitting changes, I think It would be good to include this partial speedup but I don’t mind holding it off in a different branch if we want to wait for further improvements.

Reply:
OK, I'm happy to go with what you think is best - if you can't think of any use case for `evolve` without operators then I'm surely not going to be able to. Especially if it's currently broken, no tests, and has been "left behind" by other past changes. If it was needed in the future it would need to be rewritten anyway, and hopefully since this discussion is on Git/the code is in the commit history it would be easier than starting from scratch. I suppose we'll want to change the default argument, and validate that it has a length?

Regarding performance improvements and unexpected behaviour more generally, if it's not going to be feasible to fully investigation/address I think it's fine to leave it documented on the issue and focus on getting (comparatively) easy wins/fitting improvements released.

Reply:
I mean it might be useful, I would assume its more for academic use. I think for the purposes of not introducing a breaking change I will just fix it so it has the original behaviour it would have had before it was broken, but this time adding tests for them.